### PR TITLE
BLOCKED: Switch to pyelliptic for faster EC2 handling

### PR DESCRIPTION
**Pull request by jrconlin**
ecdsa is very flexible as a EC library, but CPU intensive. Switching to
a C lib based EC handler should make things faster and more CPU friendly
under loads.

### CODE REVIEW

**Comment by jrconlin** [comment]
I'm not quite sure why the 2.6 test is failing. It may be that the call out to OpenSSL is causing 'key' to be deallocated. It's clearly being defined. I know that 2.6 has reached EOL, so not sure how critical that test is.

**Comment by jrconlin** [comment]
Sigh, there's a bug in the signature handler. Will update with fix.

**Comment by jrconlin** [comment]
So, turns out that the backing library (pyelliptic) automatically runs whatever you pass it through SHA256. Obviously, that's bad. I've submitted a PR for that <https://github.com/yann2192/pyelliptic/pull/53>.

That PR blocks this one.

File: jose/jwk.py [hunk 1]
```diff
@@ -245,7 +259,191 @@ def verify(self, msg, sig):
...
+        if len(decoded) == 64:
+            return curve, None, "\04" + decoded
+        if decoded[0] == "\04":
+            return curve, None, decoded
+
+        try:
+            asn_set = decoder.decode(decoded)[0]
+        except:
+            raise JWKError("Invalid EC Key")
+        pri_key = self.bitstring_to_str(asn_set[1])
```
Comment (jcjones) [inline jose/jwk.py]: This is only applicable for the first case; in the `univ.Sequence` case you end up having to shuffle this around.

I'd recommend initializing this to `None`, too, and moving this set into the first case block.

File: jose/jwk.py [hunk 2]
```diff
@@ -245,7 +259,191 @@ def verify(self, msg, sig):
...
+        # A public key starts with a sequence
+        if isinstance(asn_set[0], univ.Sequence):
+            # confirm that the public key curve matches up
+            # with the OID pair type (Also includes the public key OID,
+            # which we ignore)
+            pcurve = self.curve_oids.get(asn_set[0][1])
+            if pcurve:
+                curve = pcurve
+            # Shuffle the key to the right return.
+            pub_key = pri_key
```
Comment (jcjones) [inline jose/jwk.py]: On the face of it, this is super alarming: setting the public value to the private value. Per the earlier comment, don't shuffle these. Basically, don't set `pri_key` to something that isn't a private key in the first place. If we aren't sure yet, let it remain `asn_set[1]`

File: jose/jwk.py [hunk 3]
```diff
@@ -245,7 +259,191 @@ def verify(self, msg, sig):
...
+            pri_key = None
+        if not curve:
+            raise JWKError("Invalid EC curve type key specified.")
```
Comment (jcjones) [inline jose/jwk.py]: Nit: It's not really that it was specified wrong, it's that the supplied data isn't one of our known formats.

File: jose/jws.py [hunk 4]
```diff
@@ -161,7 +159,8 @@ def _sign_header_and_claims(encoded_header, encoded_claims, algorithm, key_data)
     signing_input = b'.'.join([encoded_header, encoded_claims])
     try:
         key = jwk.construct(key_data, algorithm)
-        signature = key.sign(signing_input)
+        dat = signing_input.decode('utf8').encode('utf8')
```
Comment (jcjones) [inline jose/jws.py]: Nit: This could use a comment explaining why it's being round-tripped through the decoder/encoder

File: jose/jwk.py [hunk 5]
```diff
@@ -245,7 +259,191 @@ def verify(self, msg, sig):
...
+                privkey=raw_key,
+                pubkey=raw_pub,
+                hasher=sha_map[algorithm])
+            return
+        raise JWKError('Unable to parse an ECKey from key: %s' % key)
+
+    def repad(self, st):
+        """Add base64 padding back to the end of a stripped character
+        sequence
+
```
Comment (jcjones) [inline jose/jwk.py]: Nit: whitespace

File: jose/jwk.py [hunk 6]
```diff
@@ -245,7 +259,191 @@ def verify(self, msg, sig):
...
+            if pcurve:
+                curve = pcurve
+            # Shuffle the key to the right return.
+            pub_key = pri_key
+            pri_key = None
+        if not curve:
+            raise JWKError("Invalid EC curve type key specified.")
+        return curve, pri_key, pub_key
+
+    def _process_jwk(self, jwk_dict, algorythm="sha256"):
```
Comment (jcjones) [inline jose/jwk.py]: Nit: Spelling (Algorithm)

**Comment by jrconlin** [comment]
@jcjones Thanks again for the r?s. Addressed (and did some flake8 fixes, plus a pointer to the newer pyelliptic library patch)

File: jose/jwk.py [hunk 7]
```diff
@@ -245,7 +259,190 @@ def verify(self, msg, sig):
...
+        if (isinstance(asn_set[0], univ.Integer) and
+                asn_set[0] == 1):
+            # Followed by the OID
+            curve = self.curve_oids.get(asn_set[2])
+            if curve:
+                pri_key = self.bitstring_to_str(asn_set[1])
+                # And finally the public key
+                pub_key = self.bitstring_to_str(asn_set[3])
+        # A public key starts with a sequence
+        if isinstance(asn_set[0], univ.Sequence):
```
Comment (pjenvey) [inline jose/jwk.py]: an elif instead of if would be a bit clearer here

**Comment by xoviat** [comment]
Cryptography will be included as a dependency if I succeed, and already has [elliptic curve cryptography](https://cryptography.io/en/latest/hazmat/primitives/asymmetric/ec/) based on openssl. I'm not going to work on this, but the point is that if you use cryptography, then this won't be blocked.

**Comment by jrconlin** [comment]
So, turns out that OpenSSL (and by default, python cryptography) use a DER formatted signature value, where `ecdsa` (and a number of others) use straight raw key data. Both contain the same info, but are incompatible. I'm working on the fix, I've filed different bugs pointing out the problem, and I'm going to contact spec authors to get some clarification about this. 

The fix is "reasonably simple", (check the signature length, and if 64 octets, recompile the signature into a DER and hand that over to cryptography's EC verify. Needless to say, it's going to impact this patch, so I'm going to close this for now until the new patch is ready.

**Pull request closed without merging**